packages/contracts-bedrock/contracts/dispute/DisputeGameFactory.sol

```diff
@@ -0,0 +1,163 @@
+        Claim rootClaim,
+        bytes calldata extraData
+    ) external view returns (IDisputeGame _proxy) {
+        return disputeGames[getGameUUID(gameType, rootClaim, extraData)];
+    }
+
+    /**
+     * @notice Creates a new DisputeGame proxy contract.
+     * @notice If a dispute game with the given parameters already exists,
+     *         it will be returned.
+     * @param gameType The type of the DisputeGame.
+     *        Used to decide the proxy implementation.
+     * @param rootClaim The root claim of the DisputeGame.
+     * @param extraData Any extra data that should be provided
+     *        to the created dispute game.
+     * @return proxy The clone of the `DisputeGame`.
+     */
+    function create(
+        GameType gameType,
+        Claim rootClaim,
+        bytes calldata extraData
+    ) external returns (IDisputeGame proxy) {
+        // Grab the implementation contract for the given `GameType`.
+        IDisputeGame impl = gameImpls[gameType];
+
+        // If there is no implementation to clone for the given `GameType`, revert.
+        if (address(impl) == address(0)) {
+            revert NoImplementation(gameType);
+        }
+
+        // Clone the implementation contract and initialize it with the given parameters.
+        bytes memory data = abi.encodePacked(rootClaim, extraData);
+        proxy = IDisputeGame(address(impl).clone(data));
+        proxy.initialize();
+
+        // Compute the unique identifier for the dispute game.
+        Hash uuid = getGameUUID(gameType, rootClaim, extraData);
+
+        // If a dispute game with the same UUID already exists, revert.
+        if (address(disputeGames[uuid]) != address(0)) {
+            revert GameAlreadyExists(uuid);
+        }
+
+        // Store the dispute game in the mapping & emit the `DisputeGameCreated` event.
+        disputeGames[uuid] = proxy;
+        emit DisputeGameCreated(address(proxy), gameType, rootClaim);
+    }
+
+    /**
+     * @notice Sets the implementation contract for a specific `GameType`.
+     * @param gameType The type of the DisputeGame.
+     * @param impl The implementation contract for the given `GameType`.
+     */
+    function setImplementation(GameType gameType, IDisputeGame impl) external onlyOwner {
+        gameImpls[gameType] = impl;
+    }
+
+    /**
+     * @notice Returns a unique identifier for the given dispute game parameters.
+     * @dev Hashes the concatenation of `gameType . rootClaim . extraData`
+     *      without expanding memory.
+     * @param gameType The type of the DisputeGame.
+     * @param rootClaim The root claim of the DisputeGame.
+     * @param extraData Any extra data that should be provided to the created dispute game.
+     * @return _uuid The unique identifier for the given dispute game parameters.
+     */
+    function getGameUUID(
+        GameType gameType,
+        Claim rootClaim,
+        bytes memory extraData
+    ) public pure returns (Hash _uuid) {
+        assembly {
```

Comment:
This isn't @refcell's fault, but solhint not allowing `assembly ("memory-safe")` is so sad

Reply:
++